src/compiler/checker.ts

```diff
@@ -32946,6 +32946,10 @@ namespace ts {
                 case SyntaxKind.ExclamationEqualsToken:
                 case SyntaxKind.EqualsEqualsEqualsToken:
                 case SyntaxKind.ExclamationEqualsEqualsToken:
+                    if (isLiteralExpressionOfObject(left) || isLiteralExpressionOfObject(right)) {
+                        const eqType = operator === SyntaxKind.EqualsEqualsToken || operator === SyntaxKind.EqualsEqualsEqualsToken;
+                        error(errorNode, Diagnostics.This_condition_will_always_return_0_since_JavaScript_compares_objects_by_reference_not_value, eqType ? "false" : "true");

```

Comment:
I think this is going to double-report since `reportOperatorErrorUnless` isn't wrapped in an `else`. Can you add a test like 

```ts
const b = [1]
if ({ a : 1 } == b) { }
```

to make sure that it doesn't?

Reply:
Yes, it will double report. Is that a problem?

```js
    if ({ a: 1 } === b) {
        ~~~~~~~~~~~~~~
!!! error TS2367: This condition will always return 'false' since the types '{ a: number; }' and 'number[]' have no overlap.
        ~~~~~~~~~~~~~~
!!! error TS2823: This condition will always return 'false' since JavaScript compares objects by reference, not value.
    }
```

Reply:
Not a big problem. At the time I thought it was going to be a duplicate of the same error.